### PR TITLE
#0: refactor ttnn device operation code and program cache

### PR DESCRIPTION
**Pull request by arakhmati**
### Ticket
n/a

### Problem description
We currently have 2 ways of writing and running a device operation. It will take time to re-write from the old way to the new way. However, they can at least be launched the same way by wrapping old device operations into new one. Additionally, new device operations don't disable program cache, but it is a requested feature.

### What's changed
- Wrapped old device operation infra using new infra
- Updated new device op infra to allow disabling program cache
- Deleted HostOperation class completely

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt_metal/tools/profiler/op_profiler.hpp
Comment (mo-tenstorrent) [inline tt_metal/tools/profiler/op_profiler.hpp:1]: ❤️

File: tt_metal/tools/profiler/op_profiler.hpp [hunk 1]
```diff
@@ -459,9 +379,7 @@ inline std::string op_meta_data_serialized_json(
 
 #else
 
-#define TracyOpTTNNDevice( \
-    op_id, op_hash, is_cached, device_id, operation, program, input_tensors, optional_input_tensors, output_tensors)
-#define TracyOpTNNNDeviceV2( \
+#define TracyOpTNNNDevice( \
```
Comment (mo-tenstorrent) [inline tt_metal/tools/profiler/op_profiler.hpp:382]: Extra N?

Comment (arakhmati) [reply tt_metal/tools/profiler/op_profiler.hpp]: fixed it

**Review by mo-tenstorrent — APPROVED** [review]

**Review by eyonland — APPROVED** [review]

**Review by tt-aho — APPROVED** [review]

File: ttnn/cpp/ttnn/device_operation.hpp [hunk 2]
```diff
@@ -253,73 +256,242 @@ inline void log_operation(
...
+    const auto& storage = std::get<tt::tt_metal::MultiDeviceStorage>(first_tensor.get_storage());
+    using storage_t = std::remove_cvref_t<decltype(storage)>;
+
+    auto num_shards = storage.num_buffers();
+
+    std::vector<tensor_return_value_t> outputs;
+    outputs.reserve(num_shards);
+
+    bool launch_shards_in_parallel = false;
+    if (launch_shards_in_parallel) {
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/device_operation.hpp:435]: @arakhmati why half implemented feature was merged together with refactoring?

**Pull request merged**